mat/format_value.go

```diff
@@ -0,0 +1,143 @@
+// Copyright ©2021 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package mat
+
+import (
+	"fmt"
+	"strconv"
+	"sync"
+)
+
+const (
+	syntaxDefault = iota
+	syntaxMATLAB
+	syntaxPython
+
+	defaultDotByte = '.'
+)
+
+// valueFormatter is a type that can set options for formatting values and that
+// satisfies the fmt.Formatter interface.
+type valueFormatter interface {
+	fmt.Formatter
+
+	setDotByte(b byte)
+	setSyntax(s int)
+}
+
+// vFormatOption is a functional option for value formatting.
+type vFormatOption func(valueFormatter)
+
+// vDotByte sets the dot character to b. The dot character is used to replace a
+// value if the result is printed with the fmt ' ' verb flag. Without a DotByte
+// option, the default dot character is '.'.
+func vDotByte(b byte) vFormatOption {
+	return func(f valueFormatter) { f.setDotByte(b) }
+}
+
+// vFormatMATLAB sets the printing behavior to output MATLAB syntax. If MATLAB
+// syntax is specified, the ' ' verb flag is ignored.
+func vFormatMATLAB() vFormatOption {
+	return func(f valueFormatter) { f.setSyntax(syntaxMATLAB) }
+}
+
+// vFormatPython sets the printing behavior to output Python syntax. If Python
+// syntax is specified, the ' ' verb flag is ignored.
+func vFormatPython() vFormatOption {
+	return func(f valueFormatter) { f.setSyntax(syntaxPython) }
+}
+
+// formattedFloat returns a fmt.Formatter for the floating point value v using
+// the given options.
+func formattedFloat(v float64, options ...vFormatOption) fmt.Formatter {
+	f := floatFormatter{
+		value:  v,
+		buf:    make([]byte, 0, 64),
+		dot:    defaultDotByte,
+		syntax: syntaxDefault,
+	}
+	for _, o := range options {
+		o(&f)
+	}
+	return &f
+}
+
+// floatFormatter formats 64-bit a floating point value and satifies the
+// valueFormatter interface. A floatFormatter utilizes an internal buffer while
+// generating formatting output and this buffer persists with the formatter.
+// For this reason, and because a floatFormatter might be reused or revalued,
+// the floatFormatter embeds an exclusion lock.
```

Comment:
Reuse doesn't require a lock, only if there is concurrent use, but concurrent use it not a goal, so the mutex can be dropped.